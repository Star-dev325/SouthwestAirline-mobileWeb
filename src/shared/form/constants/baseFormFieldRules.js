import validator from 'src/shared/form/formValidators/validator';
import i18n from '@swa-ui/locale';

const isRequired = true;

export const baseFieldRules = {
  addressLine1: [
    {
      isRequired
    },
    {
      msg: i18n('SHARED__ERROR_MESSAGES__ADDRESS_LINE_1'),
      validator: validator.isAddress
    },
    {
      msg: i18n('SHARED__ERROR_MESSAGES__ADDRESS_LINE_1_LENGTH'),
      validator: validator.isLengthLessOrEqual(40)
    }
  ],
  addressLine2: [
    {
      msg: i18n('SHARED__ERROR_MESSAGES__ADDRESS_LINE_2'),
      validator: validator.isAddress
    },
    {
      msg: i18n('SHARED__ERROR_MESSAGES__ADDRESS_LINE_2_LENGTH'),
      validator: validator.isLengthLessOrEqual(40)
    }
  ],
  city: [
    {
      isRequired
    },
    {
      msg: i18n('SHARED__ERROR_MESSAGES__CITY'),
      validator: validator.isCity
    }
  ]
};

export const basePostal = {
  stateProvince: [
    {
      msg: i18n('SHARED__ERROR_MESSAGES__STATE_PROVINCE_REGION'),
      validator: validator.isStateProvinceRegion
    },
    {
      msg: i18n('SHARED__ERROR_MESSAGES__STATE_PROVINCE_REGION_LENGTH'),
      validator: validator.isLengthLessThan(50)
    }
  ],
  isUSPostal: [
    {
      msg: i18n('SHARED__ERROR_MESSAGES__ZIP_FOR_US'),
      validator: (value) => validator.isNumeric(value) && validator.isLengthEql(5)(value)
    }
  ],
  postalCode: [
    {
      msg: i18n('SHARED__ERROR_MESSAGES__POSTAL_CODE'),
      validator: validator.isPostalCode
    },
    {
      msg: i18n('SHARED__ERROR_MESSAGES__POSTAL_CODE_LENGTH'),
      validator: validator.isLengthLessOrEqual(10)
    }
  ]
};
