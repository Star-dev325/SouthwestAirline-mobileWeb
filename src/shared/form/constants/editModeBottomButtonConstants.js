export default {
  PRIMARY: 'PRIMARY',
  UPDATE: 'UPDATE',
  DELETE: 'DELETE',
  buttonText: {
    PRIMARY: 'Primary',
    UPDATE: 'Update',
    DELETE: 'Delete'
  }
};
