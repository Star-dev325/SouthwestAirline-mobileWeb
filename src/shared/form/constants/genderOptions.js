export default {
  Male: 'Male',
  Female: 'Female'
};
