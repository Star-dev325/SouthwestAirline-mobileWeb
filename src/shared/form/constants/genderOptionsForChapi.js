export default {
  M: 'Male',
  F: 'Female'
};
