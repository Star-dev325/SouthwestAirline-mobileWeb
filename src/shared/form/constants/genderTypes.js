export default {
  FEMALE: 'F',
  MALE: 'M',
  MALE_FULL: 'MALE',
  FEMALE_FULL: 'FEMALE',
  UNAVAILABLE: 'UNAVAILABLE'
};
