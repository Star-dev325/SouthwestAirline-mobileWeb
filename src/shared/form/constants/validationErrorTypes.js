export const REQUIRED_ERROR = 'REQUIRED_ERROR';
export const FIELD_ERROR_MESSAGE = 'FIELD_ERROR_MESSAGE';
export const ERROR_HEADER = 'ERROR_HEADER';
export const SIMPLE_ERROR_POPUP = 'SIMPLE_ERROR_POPUP';
export const API_ERROR_POPUP = 'API_ERROR_POPUP';

export const DEFAULT_ERROR_TYPE = FIELD_ERROR_MESSAGE;
