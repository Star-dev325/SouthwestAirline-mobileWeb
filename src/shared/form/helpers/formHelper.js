// @flow
const repeatStr = (str, times = 1) => {
  let result = str;

  while (times > 1) {
    result += str;
    times--;
  }

  return result;
};

export const getMaskProps = function (mask: string | { rule: string, repeat?: number, maskChar?: string }) {
  const defaultMask = {
    repeat: 1,
    maskChar: null
  };

  const maskConfigs = Object.assign({}, defaultMask, typeof mask === 'string' ? { rule: mask } : mask);

  const finalMaskStr = repeatStr(maskConfigs.rule, maskConfigs.repeat);

  return {
    mask: finalMaskStr,
    maskChar: maskConfigs.maskChar,
    maxLength: finalMaskStr.length // mask can't work properly in some android device so use maxLength to limit the user input, see bug#2026
  };
};
