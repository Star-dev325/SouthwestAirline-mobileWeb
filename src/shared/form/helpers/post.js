import _ from 'lodash';
import BrowserObject from 'src/shared/helpers/browserObject';

const { document } = BrowserObject;

const post = (formProps, params) => {
  const createForm = () => {
    const form = document.createElement('form');
    const { target } = formProps;

    form.setAttribute('method', formProps.method || 'POST');
    form.setAttribute('action', formProps.path);

    if (target) {
      form.setAttribute('target', target);
    }

    return form;
  };

  const createSubmitButton = () => {
    const submitButton = document.createElement('input');

    submitButton.setAttribute('type', 'submit');
    submitButton.setAttribute('style', 'display: none;');
    submitButton.setAttribute('value', 'submitButton');

    return submitButton;
  };

  const createInput = (key, value) => {
    const input = document.createElement('input');

    input.setAttribute('type', 'hidden');
    input.setAttribute('name', key);
    input.setAttribute('value', value);

    return input;
  };

  if (formProps && formProps.path) {
    const form = createForm();

    _.each(params, (value, key) => {
      form.appendChild(createInput(key, params[key]));
    });

    form.appendChild(createSubmitButton());
    document.body.appendChild(form);
    form.submit();
    form.parentNode.removeChild(form);
  }
};

export default post;
