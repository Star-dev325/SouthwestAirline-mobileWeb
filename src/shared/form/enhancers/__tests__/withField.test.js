import { fireEvent, render } from '@testing-library/react';
import numeral from 'numeral';
import React from 'react';
import Input from 'src/shared/components/input';
import { FIELD_ERROR_MESSAGE } from 'src/shared/form/constants/validationErrorTypes';
import withField from 'src/shared/form/enhancers/withField';
import { createMockedForm, createMockedFormStore } from 'test/unit/helpers/formTestUtils';

describe('withField', () => {
  let store;
  let onSubmitStub;

  beforeEach(() => {
    onSubmitStub = jest.fn();
    store = createMockedFormStore();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  const submitForm = (container) => {
    fireEvent.submit(container.querySelector('form'));
  };

  const enterText = (container, text) => {
    fireEvent.change(container.querySelector('input'), { target: { value: text } });
  };

  describe('register and unregister', () => {
    it('submitted form data should only contain the registed fields', () => {
      const { container } = createComponent({ name: 'fieldName' });

      submitForm(container);

      expect(onSubmitStub).toHaveBeenCalledWith({
        fieldName: '100.00'
      });
    });
  });

  describe('format and parse', () => {
    it('should format the value when pass it to inputs', () => {
      const mockFormat = jest.fn().mockReturnValue(100);

      createComponent(
        { name: 'fieldName' },
        { format: mockFormat }
      );

      expect(mockFormat).toHaveBeenCalledWith('100.00');
    });

    it('should parse the value when pass it to form', () => {
      const mockFormat = jest.fn().mockReturnValue(100);
      const mockParse = jest.fn().mockReturnValue(200);
      const { container } = createComponent(
        { name: 'fieldName' },
        {
          format: mockFormat,
          parse: mockParse
        }
      );

      enterText(container, { target: { value: '200' } });
      submitForm(container);

      expect(onSubmitStub).toHaveBeenCalledWith({ fieldName: 200 });
    });
  });

  describe('validation errors', () => {
    it('should render an error message', () => {
      const { container } = createComponent(
        { name: 'fieldName' },
        { parse: (event) => event.target.value }
      );

      enterText(container, 'Some invalid text');
      submitForm(container);

      expect(container).toMatchSnapshot();
    });

    it('should clear error message when update the field value', () => {
      const { container } = createComponent(
        { name: 'fieldName' },
        { parse: (event) => event.target.value }
      );

      enterText(container, 'Some invalid text');
      submitForm(container);
      enterText(container, '100');

      expect(container).toMatchSnapshot();
    });
  });

  describe('onChange', () => {
    it('should trigger onChange event when the field component has onChange prop and value changed', () => {
      const onChangeStub = jest.fn();
      const { container } = createComponent(
        {
          name: 'field',
          onChange: onChangeStub
        },
        {
          parse: (event) => numeral(event.target.value).format('0.00')
        }
      );

      enterText(container, '200');

      expect(onChangeStub).toHaveBeenCalledWith('200.00');
    });
  });

  function createComponent(fieldProps = {}, options = {}) {
    const MockedInput = (props) => <Input {...props} />;
    const ConnectedField = withField(options)(MockedInput);

    const isNumber = (value) => typeof numeral(value).value() === 'number';
    const MockedForm = createMockedForm(store, {
      formValidator: () => (formData) =>
        (isNumber(formData.fieldName) ? {} : { fieldName: { type: FIELD_ERROR_MESSAGE, msg: 'Field must be a number' } })
    });

    return render(
      <MockedForm onSubmit={onSubmitStub} initialFormData={{ fieldName: '100.00' }}>
        <ConnectedField {...fieldProps} />
      </MockedForm>
    );
  }
});
