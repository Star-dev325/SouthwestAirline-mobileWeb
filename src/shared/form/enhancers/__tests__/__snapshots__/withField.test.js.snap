// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withField validation errors should clear error message when update the field value 1`] = `
<div>
  <form>
    <div
      class="field"
    >
      <div
        class="input"
      >
        <input
          class=""
          name="fieldName"
          value="100"
        />
      </div>
    </div>
  </form>
</div>
`;

exports[`withField validation errors should render an error message 1`] = `
<div>
  <form>
    <div
      class="field error"
    >
      <div
        class="input"
      >
        <input
          class=""
          name="fieldName"
          value="Some invalid text"
        />
      </div>
      <span
        class="field--error-msg"
      >
        Field must be a number
      </span>
    </div>
  </form>
</div>
`;
